**Preamble**

**Colloboration Policy**. The student is to *explicitly identify* his/her collaborators in the assignment. If the student did not work with anyone, he/she should indicate `Collaborators=['none']`. If the student obtains a solution through research (e.g., on the web), acknowledge the source, but *write up the solution in HIS/HER OWN WORDS*. There will be a one mark penalty if a student fails to indicate his/her collaborators.

**There will be NO EXCEPTIONS to this grading policy.**

Name = 'ZHOU DUO'
Collaborators = 'NONE'

In [1]:
# Marks 
# DO NOT CHANGE THIS

parts    = ['(0) ', '(a) ', '(b) ', '(c) ', '(d) ', '(e) ']
score    = [0,0,0,0,0,0]
maxscore = [5,2,2,2,2,2]

numparts = len(parts)

for i in range(numparts):
    print("score for {}: {} out of {}".format(parts[i], score[i], maxscore[i]))
print()
print("total score   : {} out of {}".format(sum(score),sum(maxscore)))

score for (0) : 0 out of 5
score for (a) : 0 out of 2
score for (b) : 0 out of 2
score for (c) : 0 out of 2
score for (d) : 0 out of 2
score for (e) : 0 out of 2

total score   : 0 out of 15


# Assignment 2 - Sensitivity Analysis

If you need help on using Jupyter notebooks, click <a href='#help'>here</a>. 

Objective:

In this exercise, we perform *sensity analysis* to a specific linear program (LP). 
That is, we analyse the effect of changing parameters on the optimal solution and optimal value.
Specifically, we will look at the following scenarios:

(a) Change in *objective function* for the coefficient of a *nonbasic* variable.

(b) Change in *objective function* for the coefficient of a *basic* variable.

(c) Change in a *RHS* value.

(d) Adding a new *variable*.

(e) Adding a new *constraint*.



# Linear Program LP1

For this assignment, **LP1** refers to the following linear program:

$$
\begin{array}{crcrcrcrl}
\max &   &   & x_2 & - & x_3 & + & x_4\\ 
\text{subject to} 
& x_1 & - &  x_2 & - & 3x_3 & + &  x_4 & \le 7\\ 
&3x_1 & + &  x_2 & + &  x_3 & + & 2x_4 & \le 2\\ 
&6x_1 & + & 2x_2 & + & 3x_3 & - &  x_4 & \le 1\\ 
\end{array}
$$

Note that when we refer to **LP1**. It refers to above problem with **no other modifications**.

In [4]:
# WE COMPUTE THE SOLUTION FOR LP1 HERE

import pulp

model = pulp.LpProblem("LP1", pulp.LpMaximize)

x1 = pulp.LpVariable('x1', lowBound=0)
x2 = pulp.LpVariable('x2', lowBound=0)
x3 = pulp.LpVariable('x3', lowBound=0)
x4 = pulp.LpVariable('x4', lowBound=0)

model += x2-x3+x4, "Z"


model +=   x1 -   x2 - 3*x3 +   x4  <=  7,           "constraint1"
model += 3*x1 +   x2 +   x3  + 2*x4 <=  2,           "constraint2"
model += 6*x1 + 2*x2 + 3*x3 -   x4  <=  1,           "constraint3"

print(model)

model.solve()

print("Z  : {}".format(pulp.value(model.objective)))

print("x1 : {}".format(x1.varValue))
print("x2 : {}".format(x2.varValue))
print("x3 : {}".format(x3.varValue))
print("x4 : {}".format(x4.varValue))


LP1:
MAXIMIZE
1*x2 + -1*x3 + 1*x4 + 0
SUBJECT TO
constraint1: x1 - x2 - 3 x3 + x4 <= 7

constraint2: 3 x1 + x2 + x3 + 2 x4 <= 2

constraint3: 6 x1 + 2 x2 + 3 x3 - x4 <= 1

VARIABLES
x1 Continuous
x2 Continuous
x3 Continuous
x4 Continuous

Z  : 1.4
x1 : 0.0
x2 : 0.8
x3 : 0.0
x4 : 0.6


**(0) (5 marks)** Fill in the entries for the *optimal* simplex table.
Show your working. However, feel free to use `numpy` to do the computations (click <a href='#matrix'>here</a> for simple `numpy` matrix operations).

Do take note of the order of the decision variables.

---

*You may use this for working*

First, we include the slack variables and solve the LP using `PuLP`.

In [1]:
# You may need to use python for some computations

# We compute the solution for LP1 with slack variables
from pulp import *

# WRITE YOUR CODE HERE
# Objective coe
z = [0,1,-1,1,0,0,0]

# A
a = [[1,-1,-3,1,1,0,0], 
     [3,1,1,2,0,1,0],
     [6,2,3,-1,0,0,1]]

b = [7,2,1]

# define question
m = pulp.LpProblem(sense=pulp.LpMaximize)

# define x
x = [pulp.LpVariable(f'x{i}', lowBound=0) for i in [1,2,3,4,5,6,7]]

# z %*% x Matrix
m += pulp.lpDot(z, x)

# Set conditions
for i in range(len(a)):
    m += (pulp.lpDot(a[i], x) == b[i])

m.solve()

print("Status:", LpStatus[m.status])

for v in m.variables():
    print(v.name, "=", v.varValue)

print("The optimal = ", value(m.objective))


AttributeError: module 'pulp.pulp' has no attribute 'LpMaximize'

*You may use this cell for working*

In [7]:
# Using NUMPY for calculations...

import numpy as np 

# WRITE YOUR CODE HERE
list_var = [ str(v) for v in m.variables()]
list_constr = [ c for c in m.constraints]

print ("list of variables   : ", list_var)
print ("list of constraints : ", list_constr)
print()


constraint_dict = {(c, var): 0 for c in list_constr for var in list_var}
for (var, c, coeff) in m.coefficients():
    constraint_dict[(c,var)] = coeff
    
constraint_matrix = np.matrix([[constraint_dict[(c,var)] for var in list_var] for c in list_constr])
print("constraint matrix, A: ")
print(constraint_matrix)


obj_dict = { var: 0 for var in list_var}
for (var,coeff) in m.objective.items():
    obj_dict[str(var)] = coeff

first_row = np.matrix([obj_dict[var] for var in list_var])
print("objective vector, c: ")
print(first_row)

RHS = np.matrix([- m.constraints[c].constant for c in list_constr])
print("RHS , b:")
print(RHS)

cols = ['x2','x4','x5',]
cols2 = [ 'x1',  'x3', 'x6', 'x7']

print("B, {}:".format(cols))
B = np.matrix([[constraint_dict[(c,var)] for var in cols] for c in list_constr])
print(B)

print("N, {}:".format(cols2))
N = np.matrix([[constraint_dict[(c,var)] for var in cols2] for c in list_constr])
print(N)

print("cB, {}:".format(cols))
cB = np.matrix([obj_dict[var] for var in cols])
print(cB)

print("cN, {}:".format(cols))
cN = np.matrix([obj_dict[var] for var in cols2])
print(cN)

B_N = np.linalg.inv(B) * N
print('DW_l',B_N)
B_b = np.linalg.inv(B) * RHS.T
print('DW_r',B_b)
UP_l = cN - cB * np.linalg.inv(B) * N
print('UP_l',UP_l)
UP_r = -cB * np.linalg.inv(B) * RHS.T
print('UP_r',UP_r)

print('B^-1 : ')
print(np.linalg.inv(B))
print('Optimal Simplex Tableau : ')
print(np.concatenate((np.concatenate((UP_l,UP_r),axis=1),np.concatenate((B_N,B_b),axis=1)),axis=0))    



list of variables   :  ['x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7']
list of constraints :  ['_C1', '_C2', '_C3']

constraint matrix, A: 
[[ 1 -1 -3  1  1  0  0]
 [ 3  1  1  2  0  1  0]
 [ 6  2  3 -1  0  0  1]]
objective vector, c: 
[[ 0  1 -1  1  0  0  0]]
RHS , b:
[[7 2 1]]
B, ['x2', 'x4', 'x5']:
[[-1  1  1]
 [ 1  2  0]
 [ 2 -1  0]]
N, ['x1', 'x3', 'x6', 'x7']:
[[ 1 -3  0  0]
 [ 3  1  1  0]
 [ 6  3  0  1]]
cB, ['x2', 'x4', 'x5']:
[[1 1 0]]
cN, ['x2', 'x4', 'x5']:
[[ 0 -1  0  0]]
DW_l [[ 3.   1.4  0.2  0.4]
 [ 0.  -0.2  0.4 -0.2]
 [ 4.  -1.4 -0.2  0.6]]
DW_r [[0.8]
 [0.6]
 [7.2]]
UP_l [[-3.  -2.2 -0.6 -0.2]]
UP_r [[-1.4]]
B^-1 : 
[[ 0.   0.2  0.4]
 [ 0.   0.4 -0.2]
 [ 1.  -0.2  0.6]]
Optimal Simplex Tableau : 
[[-3.  -2.2 -0.6 -0.2 -1.4]
 [ 3.   1.4  0.2  0.4  0.8]
 [ 0.  -0.2  0.4 -0.2  0.6]
 [ 4.  -1.4 -0.2  0.6  7.2]]


---

**Answer - Optimal Simplex Table**. Fill in the blanks.

$$ Original\ Tableau $$

| $Z$ | $x_2$ | $x_4$ | $s_1$ | $x_1$ | $x_3$ | $s_2$ | $s_3$ | RHS |
| :-: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |:---:|
| -1   |   1   |   1   |   0   |    0  |  -1   |  0    |    0  |   0 |
| 0   |   -1  |   1   |   1   |    1  |   -3  |   0.  |   0   | 0.8 |
| 0   |   1   |   2   |   0   |    3  |     1 |   1.  |   0   | 0.6 |
| 0   |   2   |  -1   |   0   |    6  |  3    |   0   |   1   | 7.2 |

$$ Optimal\ Tableau $$

| $Z$ | $x_2$ | $x_4$ | $s_1$ | $x_1$ | $x_3$ | $s_2$ | $s_3$ | RHS |
| :-: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |:---:|
| -1   |   0   |   0   |   0   |   -3  |  -2.2 |  -0.6 |  -0.2 |-1.4 |
| 0   |   1   |   0   |   0   |    3  |   1.4 |   0.2 |   0.4 | 0.8 |
| 0   |   0   |   1   |   0   |    0  |  -0.2 |   0.4 |  -0.2 | 0.6 |
| 0   |   0   |   0   |   1   |    4  |  -1.4 |  -0.2 |   0.6 | 7.2 |

---

**(a) (2 marks)**  Consider a new objective function 
$$ c_1x_1 +x_2-x_3+x_4.$$

For what values of $c_1$, will the optimal value remain at $1.4$?


In [21]:
#B, ['x2', 'x4', 'x5']:
import numpy as np

B = np.matrix([[-1,  1,  1],
            [ 1,  2,  0],
            [ 2, -1,  0]])

#N, ['x1', 'x3', 'x6', 'x7']:
N=np.array([[ 1, -3,  0,  0],
            [ 3,  1,  1,  0],
            [ 6,  3,  0,  1]])

#cB, ['x2', 'x4', 'x5']:
cB=np.array([[1, 1, 0]])

#cN, ['x1', 'x3', 'x6', 'x7']:
cN=np.array([[ 0, -1,  0,  0]])

B_N = np.linalg.inv(B) * N
print('B^-1 * N',B_N)
B_b = np.linalg.inv(B) * RHS.T
print('B^-1 * b',B_b)
UP_l = cN - cB * np.linalg.inv(B) * N
print('-cN + cB*B^-1*N',UP_l)
UP_l0 = cB * np.linalg.inv(B) * N
print('-cB*B^-1*N',UP_l0)
UP_r = -cB * np.linalg.inv(B) * RHS.T
print('UP_r',UP_r)

'''

B^-1 : 
[[ 0.   0.2  0.4]
 [ 0.   0.4 -0.2]
 [ 1.  -0.2  0.6]]

'''

B^-1 * N [[ 3.   1.4  0.2  0.4]
 [ 0.  -0.2  0.4 -0.2]
 [ 4.  -1.4 -0.2  0.6]]
B^-1 * b [[0.8]
 [0.6]
 [7.2]]
-cN + cB*B^-1*N [[-3.  -2.2 -0.6 -0.2]]
-cB*B^-1*N [[3.  1.2 0.6 0.2]]
UP_r [[-1.4]]


'\n\nB^-1 : \n[[ 0.   0.2  0.4]\n [ 0.   0.4 -0.2]\n [ 1.  -0.2  0.6]]\n\n'

**Write your answer to (a) here.**

$$ -c_B = \begin{bmatrix}
1 & 1 & 0  \\
\end{bmatrix}  $$

$$ -c_N=\begin{bmatrix}
0 & -1 & 0 & 0 \\
\end{bmatrix}\rightarrow \begin{bmatrix}
c_1 & -1 & 0 & 0 \\
\end{bmatrix} $$

$$ N = \begin{bmatrix}
1 & -3 & 0 & 0 \\
3 & 1 & 1 & 0 \\
6  & 3 & 0 & 1 \\
\end{bmatrix}, B = \begin{bmatrix}
-1 & 1 & 1 \\
1 & 2 & 0 \\
2  & -1 & 0 \\
\end{bmatrix} $$

$$ B^{-1}N = \begin{bmatrix}
3&   1.4&  0.2&  0.4 \\
0&  -0.2&  0.4& -0.2 \\
4&  -1.4& -0.2&  0.6 \\
\end{bmatrix},B^{-1}b = \begin{bmatrix}
0.8 \\
0.6 \\
7.2 \\
\end{bmatrix}
$$


$$ Optimal\ Tableau $$

| $Z$ | $x_2$ | $x_4$ | $s_1$ | $x_1$ | $x_3$ | $s_2$ | $s_3$ | RHS |
| :-: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |:---:|
| -1   |   0   |   0   |   0  |$c_1-3$|  -2.2 |  -0.6 |  -0.2 |-1.4 |
| 0   |   1   |   0   |   0   |    3  |   1.4 |   0.2 |   0.4 | 0.8 |
| 0   |   0   |   1   |   0   |    0  |  -0.2 |   0.4 |  -0.2 | 0.6 |
| 0   |   0   |   0   |   1   |    4  |  -1.4 |  -0.2 |   0.6 | 7.2 |



$$ c_1>3,change $$
$$ c_1\leq 3,no\ change $$

**(b) (2 marks)**  Consider a new objective function 
$$c_2x_2-x_3+x_4.$$

For what values of $c_2$, will the optimal basis remain unchanged?

---

**Write your answer to (b) here.**

$x_2=1$ change to $x_2=c_2$

$$ b= \begin{bmatrix}
7 \\ 2 \\ 1 \end{bmatrix}$$

$$ -c_B = \begin{bmatrix}
1 & 1 & 0  \\
\end{bmatrix} \rightarrow \begin{bmatrix}
c_2 & 1 & 0 \\
\end{bmatrix} $$

$$ -c_N=\begin{bmatrix}
0 & -1 & 0 & 0 \\
\end{bmatrix} $$

$$ N = \begin{bmatrix}
1 & -3 & 0 & 0 \\
3 & 1 & 1 & 0 \\
6  & 3 & 0 & 1 \\
\end{bmatrix}, B = \begin{bmatrix}
-1 & 1 & 1 \\
1 & 2 & 0 \\
2  & -1 & 0 \\
\end{bmatrix} $$

$$ B^{-1}N = \begin{bmatrix}
3&   1.4&  0.2&  0.4 \\
0&  -0.2&  0.4& -0.2 \\
4&  -1.4& -0.2&  0.6 \\
\end{bmatrix},B^{-1}b = \begin{bmatrix}
0.8 \\
0.6 \\
7.2 \\
\end{bmatrix}
$$

$$ -c_N + c_B\cdot B^{-1}N=\begin{bmatrix}
-3c_2 & -1.4c_2-0.8 & -0.2c_2-0.4 & -0.4c_2+0.2  \\
\end{bmatrix},c_B\cdot B^{-1}b=\begin{bmatrix}
-0.8c_2-0.6 \\
\end{bmatrix} $$

$$ c_2 \in \begin{cases}
-3c_2\leq 0 \\ -1.4c_2-0.8\leq 0 \\ -0.2c_2-0.4\leq 0 \\ -0.4c_2+0.2\leq 0 \\
\end{cases},\ c_2\geq \frac{1}{2} $$

---

**(c) (2 marks)**  Suppose we modify `"constraint 1"` to  
$$ x_1 - x_2 - 3x_3 + x_4 \le b_1.$$

For what values of $b_1$, will the optimal basis remain unchanged?

---

**Write your answer to (c) here.**

$$ b= \begin{bmatrix}
7 \\ 2 \\ 1 \end{bmatrix}\rightarrow  b= \begin{bmatrix}
b_1 \\ 2 \\ 1 \end{bmatrix} $$

$$  B^{-1} = \begin{bmatrix}
0&   0.2&  0.4\\
0&   0.4& -0.2\\
1&  -0.2&  0.6\\
\end{bmatrix} $$

$$ B^{-1}b = \begin{bmatrix}
0.8 \\
0.6 \\
b_1+0.2 \\
\end{bmatrix},\ c_N\cdot B^{-1}b = \begin{bmatrix}
-1.4 \\
\end{bmatrix}
$$ 

$$b_1\geq -0.2$$


---

**(d) (2 marks)** Suppose we add a new variable to **LP1** and the new linear program is:

$$
\begin{array}{crcrcrcrcrl}
\max &   &   & x_2 & - & x_3 & + & x_4 & + &c_5x_5\\ 
\text{subject to} 
& x_1 & - &  x_2 & - & 3x_3 & + &  x_4 & + & x_5 & \le 7\\ 
&3x_1 & + &  x_2 & + &  x_3 & + & 2x_4 & + & x_5 & \le 2\\ 
&6x_1 & + & 2x_2 & + & 3x_3 & - &  x_4 & + & x_5 & \le 1\\ 
\end{array}
$$
For what values of $c_5$, will the optimal value remain unchanged?


In [28]:
#B, ['x2', 'x4', 'x6']:
import numpy as np

B = np.matrix([[-1,  1,  1],
            [ 1,  2,  0],
            [ 2, -1,  0]])

#N, ['x1', 'x3', 'x5', 'x7','x8']:
N=np.array([[ 1, -3,1,  0,  0],
            [ 3,  1,1,  1,  0],
            [ 6,  3,1,  0,  1]])

#cB, ['x2', 'x4', 'x6']:
cB=np.array([[1, 1, 0]])

#cN, ['x1', 'x3', 'x5', 'x7','x8']:
cN=np.array([[ 0, -1,1,  0,  0]])

B_N = np.linalg.inv(B) * N
print('B^-1 * N',B_N)
B_b = np.linalg.inv(B) * RHS.T
print('B^-1 * b',B_b)
UP_l = cN - cB * np.linalg.inv(B) * N
print('-cN + cB*B^-1*N',UP_l)
UP_l0 = cB * np.linalg.inv(B) * N
print('-cB*B^-1*N',UP_l0)
UP_r = -cB * np.linalg.inv(B) * RHS.T
print('UP_r',UP_r)

'''

B^-1 : 
[[ 0.   0.2  0.4]
 [ 0.   0.4 -0.2]
 [ 1.  -0.2  0.6]]

'''

B^-1 * N [[ 3.   1.4  0.6  0.2  0.4]
 [ 0.  -0.2  0.2  0.4 -0.2]
 [ 4.  -1.4  1.4 -0.2  0.6]]
B^-1 * b [[0.8]
 [0.6]
 [7.2]]
-cN + cB*B^-1*N [[-3.  -2.2  0.2 -0.6 -0.2]]
-cB*B^-1*N [[3.  1.2 0.8 0.6 0.2]]
UP_r [[-1.4]]


'\n\nB^-1 : \n[[ 0.   0.2  0.4]\n [ 0.   0.4 -0.2]\n [ 1.  -0.2  0.6]]\n\n'

---

**Write your answer to (d) here.**


$$ b= \begin{bmatrix}
7 \\ 2 \\ 1 \end{bmatrix}$$

$$ -c_B = \begin{bmatrix}
1 & 1 & 0  \\
\end{bmatrix} $$

$$ -c_N=\begin{bmatrix}
0 & -1 & 0 & 0 \\
\end{bmatrix}\rightarrow -c_N=\begin{bmatrix}
0 & -1 & c_5&0 & 0 \\
\end{bmatrix}  $$

$$ N = \begin{bmatrix}
1 & -3 & 0 & 0 \\
3 & 1 & 1 & 0 \\
6  & 3 & 0 & 1 \\
\end{bmatrix}\rightarrow 
 N = \begin{bmatrix}
1 & -3&1 & 0 & 0 \\
3 & 1&1 & 1 & 0 \\
6  & 3&1 & 0 & 1 \\
\end{bmatrix} $$

$$B = \begin{bmatrix}
-1 & 1 & 1 \\
1 & 2 & 0 \\
2  & -1 & 0 \\
\end{bmatrix} $$

$$ B^{-1}N = \begin{bmatrix}
3&   1.4&0.6&  0.2&  0.4 \\
0&  -0.2&0.2&  0.4& -0.2 \\
4&  -1.4&1.4& -0.2&  0.6 \\
\end{bmatrix},B^{-1}b = \begin{bmatrix}
0.8 \\
0.6 \\
7.2 \\
\end{bmatrix}
$$

$$ -c_N + c_B\cdot B^{-1}N=\begin{bmatrix}
-3 & -2.2 & c_5-0.8 & -0.6&-.2  \\
\end{bmatrix},c_B\cdot B^{-1}b=\begin{bmatrix}
-1.4 \\
\end{bmatrix} $$

$$ c_5\leq 0.8 $$

---

**(e) (2 marks)** Suppose we add a new constraint to **LP1** and the new constraint is:

$$
x_1+x_2+x_3+x_4\le b_4
$$
For what values of $b_4$, will the optimal solution remain unchanged?

In [60]:
#B, ['x2', 'x4', 'x6']:
import numpy as np

RHS = np.matrix([[7,2,1,10]])

B = np.matrix([[-1,  1,  1,0],
            [ 1,  2,  0,1],
            [ 2, -1,  0,1],
            [1,1,0,0]])
print()

#N, ['x1', 'x3', 'x5', 'x7','x8']:
N=np.array([[ 1, -3,  0,  0],
            [ 3,  1,  0,  0],
            [ 6,  3,  1,  0],
            [ 1,  1,  0,  1]])

#cB, ['x2', 'x4', 'x6']:
cB=np.array([[1, 1, 0, 0]])

#cN, ['x1', 'x3', 'x5', 'x7','x8']:
cN=np.array([[ 0, -1,  0,  0]])

B_N = np.linalg.inv(B) * N
print('B^-1 * N',B_N)
B_b = np.linalg.inv(B) * RHS.T
print('B^-1 * b',B_b)
B_b0 = np.linalg.inv(B)
print('B^-1',B_b0)
UP_l = cN - cB * np.linalg.inv(B) * N
print('-cN + cB*B^-1*N',UP_l)
UP_l0 = cB * np.linalg.inv(B) * N
print('-cB*B^-1*N',UP_l0)
UP_r = -cB * np.linalg.inv(B) * RHS.T
print('UP_r',UP_r)


B^-1 * N [[ 1.5   1.25  0.25  0.75]
 [-0.5  -0.25 -0.25  0.25]
 [ 3.   -1.5   0.5   0.5 ]
 [ 2.5   0.25  0.25 -1.25]]
B^-1 * b [[  7.25]
 [  2.75]
 [ 11.5 ]
 [-10.75]]
B^-1 [[ 0.   -0.25  0.25  0.75]
 [ 0.    0.25 -0.25  0.25]
 [ 1.   -0.5   0.5   0.5 ]
 [-0.    0.75  0.25 -1.25]]
-cN + cB*B^-1*N [[-1. -2.  0. -1.]]
-cB*B^-1*N [[1. 1. 0. 1.]]
UP_r [[-10.]]


---

**Write your answer to (e) here.**

$$
\begin{array}{crcrcrcrcrl}
\max &   &   & x_2 & - & x_3 & + & x_4 & \\ 
\text{subject to} 
& x_1 & - &  x_2 & - & 3x_3 & + &  x_4 & \le 7\\ 
&3x_1 & + &  x_2 & + &  x_3 & + & 2x_4 & \le 2\\ 
&6x_1 & + & 2x_2 & + & 3x_3 & - &  x_4 & \le 1\\ 
&x_1&+&x_2&+&x_3&+&x_4&\le b_4
\end{array}
$$


$$ b= \begin{bmatrix}
7 \\ 2 \\ 1 \end{bmatrix} \rightarrow b= \begin{bmatrix}
7 \\ 2 \\ 1 \\ b_4 \end{bmatrix} $$

$$ -c_B = \begin{bmatrix}
1 & 1 & 0 \\
\end{bmatrix}\rightarrow -c_B = \begin{bmatrix}
1 & 1 & 0 &0 \\
\end{bmatrix}$$

$$ -c_N=\begin{bmatrix}
0 & -1 & 0 & 0 \\
\end{bmatrix}$$

$$ N = \begin{bmatrix}
1 & -3 & 0 & 0 \\
3 & 1 & 1 & 0 \\
6  & 3 & 0 & 1 \\
\end{bmatrix}\rightarrow
 N = \begin{bmatrix}
1 & -3 & 0 & 0 \\
3 & 1 & 0 & 0 \\
6  & 3 & 1 & 0 \\
1  & 1 & 0 & 1 \\
\end{bmatrix} $$

$$B = \begin{bmatrix}
-1 & 1 & 1 \\
1 & 2 & 0 \\
2  & -1 & 0 \\
\end{bmatrix}\rightarrow B = \begin{bmatrix}
-1 & 1 &1& 0 \\
1 & 2 &0& 1 \\
2  & -1 &0& 0 \\
1  & 1 &0& 0 \\
\end{bmatrix} \rightarrow B^{-1} = \begin{bmatrix}
0 & -0.25 &0.25& 0.75 \\
0 & 0.25 &-0.25& 0.25 \\
1  & -0.5 &0.5& 0.5 \\
0  & 0.75 &0.25& -1.25 \\
\end{bmatrix} $$


$$ B^{-1}N = \begin{bmatrix}
1.5&   1.25&0.25&  0.75 \\
-0.5&  -0.25&-0.25&  0.25\\
3&  -1.5&0.5&  0.5 \\
2.5&  0.25& 0.25&  -1.25 \\
\end{bmatrix},B^{-1}b = \begin{bmatrix}
-0.25+0.75b_4 \\
0.25+0.25b_4 \\
6.5+0.5b_4 \\
1.75-1.25b_4\\
\end{bmatrix}
$$

$$ -c_N + c_B\cdot B^{-1}N=\begin{bmatrix}
-1 & -2 & 0 & -1  \\
\end{bmatrix},c_B\cdot B^{-1}b=\begin{bmatrix}
-b_4 \\
\end{bmatrix} $$

$$ b_4 \in \begin{cases}
-0.25+0.75b_4\leq 0 \\ 0.25+0.25b_4\leq 0 \\ 6.5+0.5b_4\leq 0 \\ 1.75-1.25b_4\leq 0
\end{cases},\ 1.4\geq b_4\geq \frac{1}{3} $$

$$ \therefore  b_4 =  1.4 $$

---

---

In [59]:
# You may need to use python for some computations

# We compute the solution for LP1 with slack variables
from pulp import *
import pulp
# WRITE YOUR CODE HERE
# Objective coe
z = [0,1,-1,1,0,0,0,0]

# A
a = [[1,-1,-3,1,1,0,0,0], 
     [3,1,1,2,0,1,0,0],
     [6,2,3,-1,0,0,1,0],
     [1,1,1,1,0,0,0,1]]

b = [7,2,1,2]

# define question
m = pulp.LpProblem(sense=pulp.LpMaximize)

# define x
x = [pulp.LpVariable(f'x{i}', lowBound=0) for i in [1,2,3,4,5,6,7,8]]

# z %*% x Matrix
m += pulp.lpDot(z, x)

# Set conditions
for i in range(len(a)):
    m += (pulp.lpDot(a[i], x) <= b[i])

m.solve()

print("Status:", LpStatus[m.status])

for v in m.variables():
    print(v.name, "=", v.varValue)

print("The optimal = ", value(m.objective))



Status: Optimal
x1 = 0.0
x2 = 0.8
x3 = 0.0
x4 = 0.6
x5 = 0.0
x6 = 0.0
x7 = 0.0
x8 = 0.0
The optimal =  1.4


---

### Appendix

---

### Using `numpy` for matrix operations

<a id='matrix'></a>


Suppose we want to compute $$U^{-1}V$$
where
$$ 
U =\left(
\begin{array}{ccc}
1 & -6 & 0\\
0 & 2 & 1\\
0 & 2 & 0\\
\end{array}
\right)
\quad
V =\left(
\begin{array}{ccc}
1  &  3 &  0\\
16 & -3 &  0\\
11 & -1 & -1\\
\end{array}
\right)\,.
$$

The syntax is as follows.

In [3]:
import numpy as np # REMEMBER TO INCLUDE THIS LINE!

U = np.matrix([
    [1,-6,0],
    [0, 2,1],
    [0, 2,0],
]
)

V = np.matrix([
    [ 0, 3, 0],
    [16,-3, 0],
    [11,-1,-1],
]
)

M = np.linalg.inv(U)*V
print(M)

[[ 33.    0.   -3. ]
 [  5.5  -0.5  -0.5]
 [  5.   -2.    1. ]]


---

<a id='help'></a>
**Using iPython Notebooks**. When you click to the left of this box, you will notice that this box is highlighted by a slighly larger box. This is a *cell*. 

There are three types of cells in a notebook.

1. Markdown.
2. Code.
3. Raw.

You can change the type of cell by going to *Cell* on the tool bar.

You can *evaluate* cells by hitting **Shift+Enter**. Depending on the type of cells, you will have different outputs.

---

This is a **markdown** cell. Markdown is a lightweight markup language is similar to *html* with significantly less functionalities. However, the syntax is much simpler. You can find a [Markdown Cheatsheet here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

---

In [1]:
# This is a CODE cell.
# After you hit Shift+Enter, it evaluates the cell in Python.
# Take note that in Python, to comment lines, you use the symbol #

print("Hello World!")

Hello World!


**Answering Questions**. You may choose to use *raw* or *markdown* cells to answer the questions. Of course, if the answer requires you to run a routine in Python, please use a *code* cell.
